### PR TITLE
Implement title respawn module

### PR DESCRIPTION
**Pull request by Alfalik**
I'll starting working on this feature, hope some of you could help me in any different ways, what is wrong, why something is not working, what can be improved, all is welcome.

Things that does not work or are in the To-Do list:
- [x] Last hit until a player dies is not registered, so the kill/death cause will be wrong.
- [x] Titles are sent constantly only on first death, after the seconds death, they flash.
- [x] Allow players to enter regions that are denied to not-dead players.
- [x] Check if a player has left clicked.
- [x] Start a ```System.currentTimeMillis()``` on player death to store the time since a player died.
- [x] Convert ```getTimeLeftToRespawn()``` to decimals ```("0,0")```.
- [x] Prevent dead players interact with the gameplay:
 + Placing and picking up lava/water
 + Converting blocks to entities (sand and gravel)
- [x] Change the name on tab list to dark gray and bold if the player is dead, and show the player on the tab list.
- [x] Send red tinge around the player screen using world border warning distance.
- [x] The exact sound on ```PlayerDeathEvent``` & ```SnowflakeChangeEvent```.
- [x] Make the player ride an armor stand if the player is dead via packets.
- [x] Send armor stand packet if the player sneak and is dead.
- [x] Send death animation - http://wiki.vg/Protocol#Entity_Status
- [x] Check the entire drop items system.
- [x] Check if all respawn attributes work.
- [x] Allow ```message``` to have localization, as shown here:
```xml
<respawn delay="5s" spectate="true">
    <message>{translate: "death.respawn.confirmed.waiting.flagDropped"}</message>
</respawn>
```

XML documentation: https://docs.oc.tc/modules/spawns#respawn

Thanks in advance to @TheMolkaPL for the code.

### CODE REVIEW

**Comment by Pablete1234** [comment]
@Alfalik you are going to have to join the players to another team (gray) for them to change the name in the tab, since cardinal doesn't have a custom-tab implemented
also, if the player is not visible to other players, it won't show on the tablist, so maybe that is something that will need to wait until fancy-tab is implemented

**Comment by Alfalik** [comment]
I've created a ```deads``` team that deny the same events that are denied to observers. If a player dies, it's easy to force him to the ```deads``` team, but the player needs be forced back to his old team on cardinal respawn event. So, how should I store the old team of a player?

**Comment by Pablete1234** [comment]
@Alfalik the respawn module stores a list of all players dead, and when they can respawn, make another one with player + old team

**Comment by twizmwazin** [comment]
@Alfalik, I don't think it would be a good idea to keep track of dead players by switching their team. A simple hashmap in the module would probably do.

**Comment by Prodigy4532** [comment]
`src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java`

Unused Imports:

import net.md_5.bungee.api.chat.TextComponent;
import net.minecraft.server.v1_8_R3.Entity;
import org.bukkit.Statistic;
import org.bukkit.entity.Horse;
import org.bukkit.event.player.PlayerJoinEvent;

**Comment by Alfalik** [comment]
@Prodigy4532 You shouldn't care about imports, the pull request is WIP.
@twizmwazin Do you know another way to change the color name of a player in the tab list?

I've added checks on the Observers Module to deny some events that prevent the player to interact with the gameplay. There's also a HashMap that check if the player is dead or not.
And a question, what the ```message``` attribute does? It changes the title or subtitle?

**Comment by Pablete1234** [comment]
@Alfalik https://docs.oc.tc/modules/spawns
The message is displayed when the countdown got to 0 and you also clicked to respawn, but the filter that needs to be true for you to be able to respawn is false (for example KOTF maps, you can't respawn if your team has the flag), then a message is shown in the subtitle 
About how to implement it, you probably should use
```BaseComponent[] message = ComponentSerializer.parse(this.message);``` (and set it to green)

**Comment by Eudaldca** [comment]
The sound you hear when you die is `IRONGOLEM_DEATH` instead of the current one. The one you have is indeed the one others hear. Not sure if you had it in mind, but it was worth commenting. In its current state dead players count towards proximity, too.

File: src/main/java/in/twizmwaz/cardinal/module/modules/observers/ObserverModule.java [hunk 1]
```diff
@@ -152,21 +153,21 @@ public void onPlayerSpawn(CardinalSpawnEvent event) {
 
     @EventHandler
     public void onBlockChange(BlockPlaceEvent event) {
-        if (testObserver(event.getPlayer())) {
+        if (testObserver(event.getPlayer()) || TitleRespawn.isDeadUUID()) {
```
Comment (twizmwazin) [inline src/main/java/in/twizmwaz/cardinal/module/modules/observers/ObserverModule.java]: maybe just integrate this check into the testObserver method?

Comment (Alfalik) [inline src/main/java/in/twizmwaz/cardinal/module/modules/observers/ObserverModule.java]: I thought it would cause errors, I was wrong. Thank you

**Comment by Alfalik** [comment]
I've re-done some logic things about what title should be sent to the client. I've also fixed the player joining dead on ```PlayerJoinEvent``` and the player being killed on ```PlayerChangeTeamEvent```. No more minecraft vanilla respawn button :p

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    public double getTimeSinceDeath(UUID id) {
+        return ((System.currentTimeMillis() - deadPlayers.get(id)) / 1000);
+    }
+
+    public String respawnAutoMessage(Player player) {
+        player.showTitle(TextComponent.fromLegacyText(ChatColor.RED + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_DEAD).getMessage(player.getLocale())));
+        player.setSubtitle(TextComponent.fromLegacyText(ChatColor.GREEN + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_AUTO, String.valueOf(getTimeLeftToRespawn(player.getPlayer().getUniqueId()))).getMessage(player.getLocale())));
+        return null;
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: why would you want to make it a string if it always returns null? make it a
```public void respawnAutoMessage(Player player) {```

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    public String respawnClickMessage(Player player) {
+        /*if (this.message != null) {
+            player.showTitle(TextComponent.fromLegacyText(ChatColor.RED + getMessage()));
+        } else {
+            Default message here
+        }*/
+        player.showTitle(TextComponent.fromLegacyText(ChatColor.RED + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_DEAD).getMessage(player.getLocale())));
+        player.setSubtitle(TextComponent.fromLegacyText(ChatColor.GREEN + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_CLICK).getMessage(player.getLocale())));
+        return null;
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: same as above

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    public String respawnScheduleMessage(Player player) {
+        player.showTitle(TextComponent.fromLegacyText(ChatColor.RED + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_DEAD).getMessage(player.getLocale())));
+        player.setSubtitle(TextComponent.fromLegacyText(ChatColor.GREEN + new LocalizedChatMessage(ChatConstant.UI_RESPAWN_SCHEDULE, String.valueOf(getTimeLeftToRespawn(player.getPlayer().getUniqueId()))).getMessage(player.getLocale())));
+        return null;
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: again

File: src/main/java/in/twizmwaz/cardinal/util/Players.java [hunk 5]
```diff
@@ -42,6 +43,8 @@ public static String getName(ServerOperator who, boolean flairs) {
             if (player.isOnline()) {
                 if (flairs) {
                     return player.getPlayer().getDisplayName();
+                } else if (TitleRespawn.isDeadUUID(player.getUniqueId())) {
+                    return ChatColor.DARK_GRAY + "" + ChatColor.BOLD + player.getPlayer().getDisplayName();
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/util/Players.java]: name in list should change to gray, but if they writte stuff on the chat it should be their regular name (team colored), you should probably only make a "setListName" or something like that, that sets their name to their flags + gray, and when they respawn get it back to what it was

**Comment by Alfalik** [comment]
@Pablete1234 Thanks for suggesting, I changed it to a public void.

About the changes, I've make the armor stand to work. If a player dies, and the ```spectate``` attribute is false, the player will ride an armor stand; and on player respawn event, the invisible armor stand will be deleted.

**Comment by dentmaged** [comment]
@Alfalik http://dentmaged.github.io/docs/sportbukkit/org/bukkit/entity/Entity.html#getLastDamageCause--

**Comment by twizmwazin** [comment]
Can't really look over this on my phone, but it would probably be best to handle the armor stands using packets, that way they can't interfere with players and projectiles.

**Comment by Pablete1234** [comment]
@twizmwazin armorstands with a "Marker:1b" tag  have a 0x0 hitbox, they can't interfere with the game at all, invisible + marker and it won't affect players at all, the only problem is players in gm3 will be able to see them

**Comment by Alfalik** [comment]
@dentmaged thank you
@twizmwazin I'll give a try using packets
@Pablete1234 Only spigot have the ```armorStand.setMarker(boolean)``` method, the only way I found was using packets (http://wiki.vg/Entities#ArmorStand)

**Comment by Pablete1234** [comment]
@Alfalik then you have 3 choices:
- Send the armor stand with packets only to the dead player
- Send with plackets to everyone that it's a marker armor stand
- PR/Issue on sportbukkit to get the marker tag introduced

**Comment by twizmwazin** [comment]
It would be best to only send the armor stand packets to the dead players. The armor stand shouldn't technically exist. While we shouldn't use this, the marker tag is not in sportbukkit because the devs are pretty lazy about keeping it up to date with upstream bukkit and craftbukkit.

**Comment by Pablete1234** [comment]
@twizmwazin the problem with sportbukkit it's that is not "the devs" it's more like "the dev"

**Comment by Pablete1234** [comment]
@alfalik also, the last time i tested this, it had problems displaying the "you died title" because you are updating it all the time, and so, the "fade in" time always starts over, you sould send the title on death event
(player.sendTitle("you died","",fade in,integer.MAX_INTEGER, 0))
then, the run method should only player.sendSubtitle, and when the player clicks to respawn, title.clear, or reset (idk the name)

**Comment by Alfalik** [comment]
@Pablete1234 Just to clarify, what ```fadeOut``` and ```fadeIn``` mean?

**Comment by twizmwazin** [comment]
@Alfalik The time it takes for the title to go from transparent to solid

**Comment by Alfalik** [comment]
A quick update:
+ I've fixed the titles flashing.
+ Now the armor stand is sent via packets using entity metadata.
+ If anyone can tell me if the sounds sent to the killer and dead player are correct, I'll be thankful.

**Comment by Pablete1234** [comment]
@Alfalik the sound is sent by the snowflake module (+ 1 snowflake, killed someone)

**Comment by dentmaged** [comment]
For the gray name, call a RankUpdateEvent and add a check for the player death. If they are dead, make the name grey.

**Comment by Pablete1234** [comment]
@dentmaged the problem is because other players don't see them, he won't show up on the tab for anyone else (the same way obs don't show up to playing players)

**Comment by Pablete1234** [comment]
if you switch teams while you are dead, the drop items module drops the helmet (team picker)
you need to add a check on the drop item module, if player is dead, don't drop the items (they were droped when he died already)

**Comment by Prodigy4532** [comment]
The first time you die on the map , the countdown does not appear

https://youtu.be/qBZWbD_QzJA

**Comment by dentmaged** [comment]
@Prodigy4532 any logs?

**Comment by Pablete1234** [comment]
@Prodigy4532 it's the same issue i posted
@dentmaged look at my coment for logs, it's the same issue

**Comment by Alfalik** [comment]
Could anyone test and tell me if the first time a player dies, the subtitles are not sent?
I've also moved the Item Drop module to the Title Respawn module, to prevent items being droped 2 times.

**Comment by Pablete1234** [comment]
@Alfalik i think it's not happening anymore, but the armor stand does not show up at all

**Comment by Alfalik** [comment]
@Pablete1234 Armor stand is disabled until it completely works, thanks for testing anyway

**Comment by Alfalik** [comment]
How do I send the red tinge around the screen using the world border warning distance? I couldn't find anything related to world borders on sportbukkit.
About the changes, the name is now gray on tab if the player is dead.

**Comment by twizmwazin** [comment]
@Alfalik This may be of help http://wiki.vg/Protocol#World_Border

I'd suggest sending them a packet when they die and another when they respawn, again using packets so it doesn't affect anything on the server itself.

**Comment by Alfalik** [comment]
@twizmwazin Done!

**Comment by TiTo-418** [comment]
http://prntscr.com/8rhb4t

**Comment by Pablete1234** [comment]
dead players can change proximity (fly closer to the monument/core)

**Comment by Pablete1234** [comment]
@Alfalik if the armor stand is created with packets, that a player shifts would not trigger a vehile exit event, because the armor stand does not exist for the server, just detect that the player is shifting and mount him again?

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        PlayerRespawnEvent respawnEvent = new PlayerRespawnEvent(player, location, bedLocation);
+        GameHandler.getGameHandler().getPlugin().getServer().getPluginManager().callEvent(respawnEvent);
+
+        if (GameHandler.getGameHandler().getMatch().getState().equals(MatchState.PLAYING) && !Teams.getTeamByPlayer(player).get().isObserver()) {
+            player.setGameMode(GameMode.SURVIVAL);
+        } else {
+            player.setGameMode(GameMode.CREATIVE);
+        }
+
+        player.hideTitle();
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: player.setTitleTimes(0, 0, 10);

**Comment by Alfalik** [comment]
@Pablete1234 I'll check if a player is sneaking, if true, I'll send him the armor stand packet.

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java (hunk 6 above)
Comment (Alfalik) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: thanks

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+        WorldBorder playerWorldBorder = nmsPlayer.world.getWorldBorder();
+        PacketPlayOutWorldBorder worldBorder = new PacketPlayOutWorldBorder(playerWorldBorder, PacketPlayOutWorldBorder.EnumWorldBorderAction.SET_WARNING_BLOCKS);
+
+        try {
+            Field field = worldBorder.getClass().getDeclaredField("b");
+            field.setAccessible(true);
+
+            //TODO Add fields - http://wiki.vg/Protocol#World_Border
+
+            field.setInt(worldBorder, warningBlocks);
```
Comment (Alfalik) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: Wonder if somebody knows what's wrong here. I don't get errors on the console, the red border just does not work.

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+
+
+
+
+    public void sendBlood(Player player) {
+        sendWorldBorderPacket(player, 1000000000);
+    }
+
+    public void removeBlood(Player player) {
+        sendWorldBorderPacket(player, 10);
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: sendWoldBorderPacket(player, GameHandler.getGameHandler().getMatchWorld().getWorldBorder().getWarningDistance());

Comment (Alfalik) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: We need to set it, not get the actual one

Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: yes, set it to whatever the actual warning distance in that world is

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+    }
+
+    protected void sendWorldBorderPacket(Player player, int warningBlocks) {
+        EntityPlayer nmsPlayer = ((CraftPlayer) player).getHandle();
+        WorldBorder playerWorldBorder = nmsPlayer.world.getWorldBorder();
+        PacketPlayOutWorldBorder worldBorder = new PacketPlayOutWorldBorder(playerWorldBorder, PacketPlayOutWorldBorder.EnumWorldBorderAction.SET_WARNING_BLOCKS);
+
+        try {
+            Field field = worldBorder.getClass().getDeclaredField("b");
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: ```Field field = worldBorder.getClass().getDeclaredField("i");```
tested and works

File: src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+        player.getInventory().clear();
+        player.getInventory().setArmorContents(null);
+    }
+
+
+
+
+
+    public void sendBlood(Player player) {
+        sendWorldBorderPacket(player, 1000000000);
```
Comment (Pablete1234) [inline src/main/java/in/twizmwaz/cardinal/module/modules/titleRespawn/TitleRespawn.java]: ```sendWorldBorderPacket(player, Integer.MAX_VALUE);```

**Comment by Pablete1234** [comment]
@Alfalik make the 3 changes i told you and you can mark the
[X] Send red tinge around the player screen using world border warning distance.
as done

**Comment by Alfalik** [comment]
@Pablete1234 Oops, thank you

**Comment by Alfalik** [comment]
How should I parse this?
```xml
<message>{translate: "death.respawn.confirmed.waiting.flagDropped"}</message>

**Comment by twizmwazin** [comment]
To be honest, in order to do this correctly, we would need to revise the entire localization system. For now, unless someone has a better option, parse it as json, and then use the translation value to hard-link it to a translatable message in cardinal.

**Comment by dentmaged** [comment]
I would do the same as twiz, but just in more detail, replace all dots (```.```) with underscores (```_```) and then use ```Strings#getTechnicalName```, so ```death.respawn.confirmed.waiting.flagDropped``` would become ```DEATH_RESPAWN_CONFIRMED_WAITING_FLAGDROPPED``` in the enum. As well as that, use ```ChatConstant.valueOf(String);``` to get an enum value.

So your code would look like this:
```java
String translate = ...; // Parse JSON
translate = translate.replace('.', '_');
translate = Strings.getTechnicalName(translate);

ChatConstant chatConstant = ChatConstant.valueOf(translate);
ChatMessage message = new LocalizedChatMessage(chatConstant); // Or wrap it in an unlocalized chat message for colour.

**Comment by Alfalik** [comment]
@twizmwazin @dentmaged Thanks for your quick and detailed responses. Back into the topic, I'll parse it as JSON, but anyway the message attribute will be disabled until CTF gets implemented (need to check if the team of a player has the flag or not).
About the updates, I'd like somebody to check if I have added the correct sounds.

**Comment by Pablete1234** [comment]
@Alfalik you should not drop away the messages just because CTF, you have to make a filter test for the player, and see if he is allowed to respawn, when CTF is implemented, it will bring it's own filters, and they should work with the system too

**Comment by Pablete1234** [comment]
@Alfalik you need to get the filter result of trying to spawn the player, and if it's false, then display the message, if it's true,  then respawn the player

**Comment by Alfalik** [comment]
@Pablete1234 https://github.com/twizmwazin/CardinalPGM/pull/892/files#diff-d6448a104804051a50db28776e2b9f2eR140

**Comment by Pablete1234** [comment]
@Alfalik 
```
<spawn kit="somekit" team="a team" filter="the filter inside here!">
   <somewhere>
</spawn>
```

**Comment by Pablete1234** [comment]
this should maybe be a issue, because i belive respawn filters are not implemented

**Comment by Pablete1234** [comment]
@Alfalik any interest into finnishing this? if not, i may give it a try

**Comment by twizmwazin** [comment]
For sending the death animation, there is actually an API method of doing this right in Bukkit

**Comment by Alfalik** [comment]
@Pablete1234 Ops, I didn't have access to a computer. I built a new PC some days ago, I'll be coding again soon. You can always help me on Skype anyway.
@twizmwazin I'll give a look to it, thank you

**Comment by Pablete1234** [comment]
@Alfalik i'll be glad to help you finnish doing this

**Comment by Alfalik** [comment]
Oops

**Comment by Alfalik** [comment]
As an update, @Pablete1234 and I have coded the armor stand part, which is fully working. About the to-do list, none of us can't find the way to send the death animation correctly (it's only sent if the player was killed using ´´´/kill´´´), hope someone can help.

**Comment by Alfalik** [comment]
I'm back at this pull request, so I have made some changes and additions such as the death animation and completely revamp the drop items system. If somebody could test this and give me a list of errors or things that I should improve, that would be very helpful and, indeed, will help to get this pull request merged soon.

**Comment by Pablete1234** [comment]
@Alfalik did you push the changes?

**Comment by Alfalik** [comment]
@Pablete1234 yes

File: src/main/resources/lang/es.xml [hunk 11]
```diff
@@ -137,6 +137,10 @@
         <team>Equipo</team>
     </misc>
     <userInterface>
+        <deathRespawnDied>Haz muerto!</deathRespawnDied>
```
Comment (TiTo-418) [inline src/main/resources/lang/es.xml]: Has*

**Comment by twizmwazin** [comment]
Is this still a WIP? Is this ready to possibly be merged?

**Comment by Pablete1234** [comment]
@twiz it is not finnished yet, he told me he has no time anymore, and he wants me to finish it

It's just missing a rewrite of the respaw  module (to properly implement respawn filters)

**Comment by Alfalik** [comment]
@twizmwazin It needs some rewrite to get it fully working, sadly, I thought I'd have the time to finish this, but I'm about to start the 2nd year of university at wednesday. I told @Pablete1234 to finish the title respawn, since he knows most of the errors.

**Comment by Pablete1234** [comment]
#988

**Pull request closed without merging**